torch/utils/checkpoint.py

```diff
@@ -0,0 +1,114 @@
+import torch
+
+
+def detach_variable(inputs):
+    if torch.is_tensor(inputs):
+        inp = inputs.detach()
+        return inp
+    elif isinstance(inputs, tuple):
+        return tuple(detach_variable(v) for v in inputs)
+    else:
+        raise RuntimeError(
+            "Only tensor or tuple of tensors is supported. Got Unsupported input type: ", type(inputs).__name__)
+
+
+class CheckpointFunction(torch.autograd.Function):
+
+    @staticmethod
+    def forward(ctx, run_function, *args):
+        ctx.run_function = run_function
+        ctx.save_for_backward(*args)
+        with torch.no_grad():
+            outputs = run_function(*args)
+        return outputs
+
+    @staticmethod
+    def backward(ctx, *args):
+        if (not torch.autograd.is_checkpoint_valid()):
+            raise RuntimeError("Checkpointing is not compatible with .grad(), please use .backward() if possible")
+        inputs = ctx.saved_tensors
+        inputs_list = detach_variable(inputs)
+        with torch.enable_grad():
+            outputs = ctx.run_function(*inputs_list)
+
+        if torch.is_tensor(outputs):
+            outputs = (outputs,)
+        torch.autograd.backward(outputs, args)
+
+        input_grads = None
+        if isinstance(inputs_list, tuple):
+            input_grads = tuple(inp.grad for inp in inputs_list)
```

Comment:
I think the variables you get from `.detach()` never require grad, so this will always be `None`

Reply:
is it still okay though to have it as a safe and explicit thing to do?

Reply:
Sorry I don't understand your question. Can you please rephrase it?

Reply:
sure, I meant do you want these lines 39-40 removed? or can we keep them as a safe thing to have?

Reply:
Oh, they are unnecessary too, since `inputs_list` will never be a `Tensor` anyway. Let's leave the first case only

Reply:
done, only have the tuple case now, tensor case is removed.